plone/app/content/tests/test_widgets.py

```diff
@@ -680,6 +680,24 @@ def testGetMimeIcon(self):
             [{"getMimeIcon": "/plone/++resource++mimetype.icons/unknown.png"}],
         )
 
+    def testScrubHtml(self):
+        from zope.schema.vocabulary import SimpleTerm
+        from zope.schema.vocabulary import SimpleVocabulary
+
+        view = VocabularyView(self.portal, self.request)
+        vocab = SimpleVocabulary(
+            [
+                SimpleTerm(
+                    token=f"term {idx} <b>",
+                    value=f"term {idx} <b>",
+                    title=f"term {idx} <b>",
+                )
+                for idx in range(3)
+            ]
+        )
+        with mock.patch.object(view, "get_vocabulary", return_value=vocab):
+            json.loads(view())
```

Comment:
Yes, let's rename the test method, insert an explanation, and do at least one basic test on the result.

```suggestion
            result = view()
        # The above values could result in invalid json if there is an error in
        # the code: the following call would give a json.decoder.JSONDecodeError.
        # See https://github.com/plone/plone.app.content/pull/288
        parsed = json.loads(result)
        self.assertEqual(parsed["results"][0]["text"], "term 0 <b></b>")
```

Reply:
I'm not 100% sure about that check. It tests for what we expect, but we actually don't care that the result is exactly like this. It could be argued that the call ideally would simply leave the original string alone, because there's no malicious html or script to scrub out. That said, I don't know whether there's a requirement that the vocabulary items are valid HTML. Maybe it would also be fine to escape something like `<b>`? For context: I encountered the problem on a site used by historians who have tags like `Bienal de São Paulo <2, 1953-1954, São Paulo>`, `Aachen <2008>`, or `Almasch <b>`, which is of course not supposed to be HTML but probably some convention in their field. Changing this to e.g. `Almasch <b></b>` is semantically wrong and probably confusing.

Reply:
Ouch, that gets tricky.

I suppose the only way that the scrub_html method can do its work (which is: discarding potentially dangerous html parts) is by first letting lxml transform it into valid html, which already brings about this change.

A test of what is in the result is still preferable.  But I suppose it could also be something like this:

```
# The original term should be there, but we don't quite care whether the harmless
# '<b>' is kept or is replaced by '<b></b>' or even '<b />'.
self.assertIn("term 0 <b", parsed["results"][0]["text"])
```

I guess your use case is very much a corner case that we can't do much about here.

Maybe it helps if you move the [lines that call scrub_html](https://github.com/plone/plone.app.content/pull/288/files#diff-8a29717d7b832a3af9002cae606011fdfd4eea37f992e4fe0db83b6e0bccc8adL262-L268) to a method in the class.  Then for this project you can overwrite just that new method. so you don't have to override the complete `__call__` method.

The method is large enough that it warrants breaking up into smaller parts.  For instance the lines with the [condition `if attributes`](https://github.com/plone/plone.app.content/pull/288/files#diff-8a29717d7b832a3af9002cae606011fdfd4eea37f992e4fe0db83b6e0bccc8adL214) could also be moved to a separate method.  But that is optional.

Reply:
OK, let's go with your suggestion above for now, and I'll try to find some time to make further improvements in the coming weeks.